Redesign/restructure the sdk

amenski opened the pull request:
- Refactor the ApiClient to a common class with single responsibility.
- Remove redundancy
- Restructure packages for better clarity
- Add phone number validator (handles safari and ethiotelecom)

amenski committed:
small refactor and restructure packages

amenski committed:
refactor BillService

amenski committed:
refactor InstitutionService

Birhan121994 commented:
The changes made in this pull request significantly improve the design, flexibility, and maintainability of the codebase. The refactor introduces best practices such as separation of concerns, input validation, and better exception handling, making the code more robust and easier to maintain. The modular approach with improved configuration management and validation ensures that the code will scale well in future development.

Great work on implementing these improvements! With a few additional enhancements around error handling and logging, the system will be even more resilient and production-ready.
Here are my suggestion for the improvements:

**Suggestion 1: Enhance Phone Number Validation Flexibility**
To make phone number validation more flexible, consider using the following regex pattern:
r"^(\+?251|251)?0?(9|7)[0-9]{8}$"
Based on the input from the user, the phone number can be parsed and the country code +251 can be added. Below is an example of how to normalize the phone number:
```python 
if (phone.startsWith('+2510')) {
    phone = phone.substring(5);
} else if (phone.startsWith('2510')) {
    phone = phone.substring(4);  // Remove the 2510 prefix
} else if (phone.startsWith('+251')) {
    phone = phone.substring(4);  // Remove the +251 prefix
} else if (phone.startsWith('251')) {
    phone = phone.substring(3);  // Remove the 251 prefix
} else if (phone.startsWith('0')) {
    phone = phone.substring(1);  // Remove leading zero if present
}
// Concatenate +251 to the cleaned phone number
cleaned_phone = '+251' + phone;
```

**Suggestion 2: Improve Error Handling for HTTP Status Codes**
While the recent changes improve exception handling for specific cases (like validation errors), additional work could be done to handle different HTTP status codes in API responses, particularly for errors such as 400 or 500. Introducing more specific exception classes (e.g., InvalidPhoneNumberException or AirtimePurchaseFailureException) would improve clarity and make error handling more targeted and precise.
**Suggestion 3: Update .gitignore Configuration**
It is recommended to modify the .gitignore file to avoid tracking build artifacts. The inclusion of the following lines:
```bash
!**/src/main/**/out/
!**/src/test/**/out/
```
would result in Git tracking those build artifacts, which is not ideal. Make sure that build outputs are excluded from version control to keep the repository clean and reduce unnecessary file tracking. Overall great work.

amenski committed:
small refactor and adds phone number validation test

amenski commented:
Hi Birhan,
Thank you for your comment and I am glad that you liked my modifications. I will try to respond to your suggestions as below:
1. **Enhance Phone Number Validation Flexibility**: Thank you for your suggestion. I found a small issue with the pattern. 
I fixed it to be clean and easy to understand. It now covers the cases:
	- +251
	- 251
	- 0
	- No prefix at all (91034567899)
	- No need to handle numbers like +2510/2510
2. **Error Handling for HTTP Status Codes:** I agree with your idea, mostly it's a design choice to have as many classes as the number of exceptions. Or centralize it to a single type with dynamic handling.

I create individual exception classes for the following cases only:
	- When exceptions need very specific handling or behavior
	- When exceptions need to carry type-specific data
	- When different exceptions need to be caught and handled differently

But, please feel free to update them.

3. Those folders are already excluded. A **commit from the past (10 months back)** introduced the build folder to be tracked. I have removed that now.

Let me know for any doubts.
BR
